_posts/2020-08-28-porting-namecoin-in-tor-browser-to-windows.md

```diff
@@ -0,0 +1,108 @@
+
+> Attempting to attach streams
+> via TC when "__LeaveStreamsUnattached" is false may cause a race between
+> Tor and the controller, as both attempt to attach streams to circuits.}
+> 
+> You can try to attachstream to a stream that
+> has already sent a connect or resolve request but hasn't succeeded
+> yet, in which case Tor will detach the stream from its current circuit
+> before proceeding with the new attach request.}
+
+Well that certainly explains what was happening.  
+Tor was opening a stream to bootstrap, and (depending on exact timing) StemNS was trying to attach it again, which caused the stream to be detached, thus killing the bootstrap.
+
+At this point, I decided to ask the Tor developers whether this was a bug in the spec or the C implementation.  
+Roger Dingledine pointed me to the exact C code that handled this, which indicated exactly how StemNS could detect this case and handle it properly.  
+Roger also indicated that the C code was correct, and that the spec was incorrect.  
+I concurred that this made sense.
+
+I was then able to modify StemNS to handle this properly by detecting whether a new stream was created by user traffic or internal bootstrap, and only attaching streams from the former.  
+Testing confirmed that the bug was fixed.  
+Great, now let's move on.
+
+## Exiting Namecoin when Tor Browser Exits
+
+The existing Namecoin support in Tor Browser relies on a Bash script that signals Electrum-NMC and StemNS to exit after Firefox exits.  
+Alas, Bash is only used as a launcher in GNU/Linux, so I needed to port this to a more cross-platform approach.
+
+Tor Browser already solves this problem for the Tor daemon: Tor needs to exit when Firefox has done so.  
+Tor Browser does this by having Firefox send the `TAKEOWNERSHIP` command to Tor, which instructs Tor to exit when Firefox closes the control port connection.  
+This inspired me to do something similar in StemNS: I added an event listener to StemNS that triggers when Tor closes the control port connection to StemNS (which will happen when Tor exits).  
+I configured the event listener to send an RPC `stop` command to Electrum-NMC (which will make Electrum-NMC exit), and then exit StemNS as well.  
+(The latter took some DuckDuckGo-fu, as it turns out that `sys.exit()` can't be called from a child thread in Python; the correct way to exit from a child thread is `os._exit()`.)
+
+Some testing revealed that this worked; I was able to remove the Bash code that terminated Electrum-NMC and StemNS, and they still exited properly.  
+Moving on...
+
+## Launching Namecoin when Tor Browser Starts
+
+Of course, the other part of the Bash launcher for Namecoin was the code that launches Electrum-NMC and StemNS.  
+For this, I ended up copying/pasting the code in Tor Launcher that launches the Tor daemon.  
```

Comment:
(Nit) This is vague. The "/pasting" makes it unclear whether you copied code from Tor Launcher into something else (what?), or copied the Bash code into Tor Launcher (how?)
> My solution was to copy the code from Tor Launcher that launches the Tor daemon.